spring-integration-core/src/main/java/org/springframework/integration/metadata/MetadataStoreReaper.java

```diff
@@ -0,0 +1,175 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.metadata;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.util.TimeZone;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+
+/**
+ * {@link Runnable} component that deletes entries from a {@link MetadataStore}.
+ * By default, it expects the value of each entry to be a timestamp, as is the default configuration when using the
+ * {@link org.springframework.integration.handler.advice.IdempotentReceiverInterceptor} and the
+ * {@link org.springframework.integration.selector.MetadataStoreSelector}. You can inject a
+ * {@link Predicate} to implement an expiration condition.
+ *
+ * @author David Turanski
+ **/
+
+public class MetadataStoreReaper implements Runnable, SmartLifecycle {
+
+	private final static Log logger = LogFactory.getLog(MetadataStoreReaper.class);
+
+	private final MetadataStore metadataStore;
+
+	private final Predicate<String> isExpired;
+
+	private volatile boolean running;
+
+	private final ReentrantLock lifecycleLock = new ReentrantLock();
+
+	private volatile int phase = 0;
+
+	private volatile boolean autoStartup = true;
+
+	/**
+	 * Remove entries from a {@link MetadataStore} based on a maximum time-to-live.
+	 * @param metadataStore the MetadataStore.
+	 * @param timeToLive the maximum time-to-live.
+	 */
+	public MetadataStoreReaper(MetadataStore metadataStore, Duration timeToLive) {
+		this(metadataStore, value -> {
+			try {
+				long timestamp = Long.valueOf(value);
+				Duration age = Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(timestamp),
+								TimeZone.getDefault().toZoneId()), LocalDateTime.now());
+				return age.compareTo(timeToLive) >= 0;
+			}
+			catch (NumberFormatException e) {
+				logger.warn(String.format("Cannot convert %s to a long.", value));
+			}
+			return false;
+
+		});
+	}
+
+	/**
+	 * Remove entries from a {@link MetadataStore} based on a {@link Predicate} used to determine the expiry condition.
+	 * @param metadataStore the MetadataStore.
+	 * @param isExpired the Predicate tested against the entry's value.
+	 */
+	public MetadataStoreReaper(MetadataStore metadataStore, Predicate<String> isExpired) {
+		Assert.notNull(metadataStore, "'metadataStore' cannot be null");
+		Assert.notNull(isExpired, "'isExpired predicate' cannot be null");
+		this.metadataStore = metadataStore;
+		this.isExpired = isExpired;
+	}
+
+	@Override
+	public void run() {
+		this.metadataStore.keySet().stream().forEach(k ->
+		{
+			if (this.isExpired.test(this.metadataStore.get(k))) {
```

Comment:
This technique is pretty interesting and I would say that it is a time to add `expireEntries(Predicate<String>)` into the `MetadataStore` contract! 

Although we might need to think about something like `entrySet` from the store at once or as a stream do not overhead the memory. 

On the other hand it should be better to rely on the native *remove-by-filter* functionality in the store, e.g. `DELETE FROM WHERE ...` for JDBC. However in this case we need to supply a converter from String to timestamp in that query per se...

Need to think more...

Reply:
I still think this is bad idea to enforce people to iterate all the entries in the framewrok's component.

I know that we have something similar in the `MessageGroupStore`, but that is a different story and has some hard-tying with an aggregator logic.

Since here we just do *remove-by-filter* that's definitely fall to the good native `DELETE` support in the target `MetadataStore` implementation.

For some exotic use-cases like yours to remove by the timestamp representation in the `value` we already exposed that `keySet()` for iteration, however I still think that it would be better to do that on the target store directly. We always can convert string to `timestamp` in the `WHERE` clause.